.multipackage/scripts/generate_api.py

```diff
@@ -0,0 +1,140 @@
+"""Generate API docs for all subpackages."""
+
+from __future__ import print_function
+import sys
+import os
+import shutil
+import time
+import argparse
+
+VERSION = "0.1.0"
+
+
+class Error(Exception):
+    """Exception that indicates an error message."""
+
+    def __init__(self, message, code=1):
+        super(Error, self).__init__(message)
+        self.message = message
+        self.exit_code = code
+
+
+def delete_with_retry(folder):
+    """Try multiple times to delete a folder.
+
+    This is required on windows because of things like:
+    https://bugs.python.org/issue15496
+    https://blogs.msdn.microsoft.com/oldnewthing/20120907-00/?p=6663/
+    https://mail.python.org/pipermail/python-dev/2013-September/128350.html
+    """
+
+    for _i in range(0, 5):
+        try:
+            if os.path.exists(folder):
+                shutil.rmtree(folder)
+
+            return
+        except:  #pylint:disable=bare-except;We do really want to catch everything
+            time.sleep(0.1)
+
+    raise Error("Could not delete directory after 5 attempts, failing")
+
+
+def build_parser():
+    """Build an argument parser."""
+
+    parser = argparse.ArgumentParser()
+    parser.add_argument("-o", "--output", required=True, help="Output directory for saving the docs")
+    parser.add_argument("-c", "--clean", action="store_true", help="Clean the output directory before starting")
+    parser.add_argument("input", default=[], nargs="+", help="The input directories to generate docs for")
+    parser.add_argument("-t", "--template", required=True, help="The template directory")
+    parser.add_argument("-r", "--remove", action="append", default=[], help="Remove these generated files")
+    parser.add_argument('--version', action='version',
+                        version='%(prog)s {version}'.format(version=VERSION))
+    return parser
+
+
+def verify_output(path):
+    """Verify that output path exists and is a directory.
+
+    Args:
+        path (str): The destination path.
+    """
+
+    if not os.path.exists(path):
+        os.mkdir(path)
```

Comment:
If the user puts in a path with nested non-created directories, I think mkdir will fail:
```
>>> os.mkdir("test/test2/test3")
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
OSError: [Errno 2] No such file or directory: 'test/test2/test3'
>>> os.getcwd()
'/Users/mrunchey/gitrepos/arch_gateway'
>>> os.mkdir("test/test2")
>>> os.mkdir("test/test2/test3")
```

Reply:
Agreed, normally I would use `os.makedirs` for that reason but before python 3 it didn't (apparently) support if the directory already existed:

https://docs.python.org/3/library/os.html#os.makedirs

I created https://github.com/iotile/multipackage/issues/11 to track this.